spec/server/listener_metrics_spec.rb

```diff
@@ -0,0 +1,169 @@
+
+          # These are some lines that we know that should be part of the output.
+          check_lines = [
+            '# TYPE apisonator_listener_response_codes counter',
+            '# HELP apisonator_listener_response_codes Response codes',
+            'apisonator_listener_response_codes{request_type="authrep",resp_code="2xx"} 2.0',
+            'apisonator_listener_response_codes{request_type="authorize",resp_code="2xx"} 1.0',
+            'apisonator_listener_response_codes{request_type="authrep",resp_code="4xx"} 1.0',
+            '# TYPE apisonator_listener_request_times_seconds histogram',
+            '# HELP apisonator_listener_request_times_seconds Request times',
+            'apisonator_listener_request_times_seconds_count{request_type="authorize"} 1.0',
+            'apisonator_listener_request_times_seconds_count{request_type="authrep"} 3.0',
+          ]
+
+          expect(check_lines.all? { |line| metrics_resp.include?(line) }).to be true
+        end
+      end
+
+      context 'when listener metrics are not enabled' do
+        before do
+          write_config_file(config_file)
+          start_puma_without_metrics(config_file, puma_port)
+        end
+
+        after do
+          stop_puma(puma_port)
+          delete_config_file(config_file)
+        end
+
+        it 'does not open the metrics port' do
+          expect { Net::HTTP.get(puma_host, metrics_endpoint, metrics_port) }
+                 .to raise_error(Errno::EADDRNOTAVAIL)
+        end
+      end
+
+      private
+
+      def write_config_file(path)
+        File.open(path, 'w+') do |f|
+          f.write("ThreeScale::Backend.configure do |config|\n"\
+              " config.listener_prometheus_metrics.enabled = ENV['CONFIG_LISTENER_PROMETHEUS_METRICS_ENABLED'].to_s == 'true'\n"\
+              " config.listener_prometheus_metrics.port = ENV['CONFIG_LISTENER_PROMETHEUS_METRICS_PORT']\n"\
+              "end\n")
+        end
+      end
+
+      def delete_config_file(path)
+        File.delete(path)
+      end
+
+      def start_puma_with_metrics(config_file, puma_port, metrics_port)
+        start_puma(config_file, true, puma_port, metrics_port)
+      end
+
+      def start_puma_without_metrics(config_file, puma_port)
+        start_puma(config_file, false, puma_port, '')
+      end
+
+      def start_puma(config_file, metrics_enabled, puma_port, metrics_port)
+        envs = {
+          LISTENER_WORKERS: 2, # To check that metrics are accurate with multiple workers
+          CONFIG_LISTENER_PROMETHEUS_METRICS_ENABLED: metrics_enabled,
+          CONFIG_LISTENER_PROMETHEUS_METRICS_PORT: metrics_port,
+          CONFIG_FILE: config_file
+        }
+        envs_str = envs.map { |env, val| "#{env}=#{val}" }.join(' ')
+
+        # Send logs to /dev/null to avoid cluttering the output
+        system("#{envs_str} bundle exec bin/3scale_backend start -p #{puma_port} > /dev/null &")
+        sleep(2) # Give it some time to be ready
```

Comment:
Ugh. This calls for some way to add this functionality in the server code, as this ad-hoc invocation is coupling the bin program :confounded:. Well, at least you can test the retval of system to be true, otherwise continuing with the sleep and the tests doesn't make sense.

Reply:
How would you do that?

Reply:
`system "...".tap { sleep 2 }` then check the return value, maybe assert it is true so that the test fails.